verification/certs.go

```diff
@@ -65,8 +170,422 @@ func getTcbInfoForHandle(c DPEClient, handle *ContextHandle) (*ContextHandle, Di
+	}
+
+	// Check TCI_CUMULATIVE against expected cumulative TCI
+	wantCumulativeTCI := computeExpectedCumulative(lastCumulativeTCI, currentTCI)
+	cumulativeTCI := tcbInfo.Fwids[1].Digest
+	if !bytes.Equal(cumulativeTCI, wantCumulativeTCI) {
+		err = fmt.Errorf("unexpected TCI_CUMULATIVE value, want %v but got %v", wantCumulativeTCI, cumulativeTCI)
+	}
+	return err
+}
+
+// Checks the FWID block's Digest.
+// FWID at index 0 has the TCI_CURRENT as digest
+// FWID at index 1 has the TCI_CUMULATIVE as digest
+// FWID array always has two digest/hashAlg blocks when "ExtendTci" is supported by DPE profile.
+func validateDiceTcbFwids(leafCertBytes []byte, expectedCurrentTci []byte) error {
+	var leafCert *x509.Certificate
+	var err error
+
+	// Check whether certificate is DER encoded.
+	if leafCert, err = x509.ParseCertificate(leafCertBytes); err != nil {
+		return err
+	}
+
+	// Get DICE information from MultiTcbInfo Extension
+	var multiTcbInfo []DiceTcbInfo
+	if multiTcbInfo, err = getMultiTcbInfo(leafCert); err != nil {
+		return err
+	}
+
+	if len(multiTcbInfo) == 0 {
+		return fmt.Errorf("certificate MutliTcbInfo is empty")
+	}
+
+	// Check current TCI value
+	currentTci := multiTcbInfo[0].Fwids[0].Digest
+	if !bytes.Equal(expectedCurrentTci, currentTci) {
+		return fmt.Errorf("current TCI value for current TCB block, want %v but got %v", expectedCurrentTci, currentTci)
+	}
+
+	// Calculate expected cumulative value
+	var defaultTci []byte
+	var hasher hash.Hash
+	if multiTcbInfo[0].Fwids[1].HashAlg.Equal(OidSHA384) {
+		hasher = sha512.New384()
+		defaultTci = make([]byte, 48)
+		hasher.Write(defaultTci)
+	} else if multiTcbInfo[0].Fwids[1].HashAlg.Equal(OidSHA256) {
+		hasher = sha256.New()
+		defaultTci = make([]byte, 32)
+	}
+
+	// Check cumulative, current TCI of last index
+	// It must have default TCI value
+	lastIndex := len(multiTcbInfo) - 1
+	if !bytes.Equal(multiTcbInfo[lastIndex].Fwids[0].Digest, defaultTci) {
+		return fmt.Errorf("current TCI value for first TCB block, want %v but got %v", defaultTci, multiTcbInfo[lastIndex].Fwids[0].Digest)
+	}
+
+	if !bytes.Equal(multiTcbInfo[lastIndex].Fwids[1].Digest, defaultTci) {
+		return fmt.Errorf("cumulative TCI value for first TCB block, want %v but got %v", defaultTci, multiTcbInfo[lastIndex].Fwids[1].Digest)
+	}
+
+	// Check cumulative, current TCI of other indices if any
+	lastCumulativeTCI := defaultTci
+	multiTcbInfo = multiTcbInfo[:lastIndex]
+
+	for i, tcbinfo := range multiTcbInfo {
+		currentTci := tcbinfo.Fwids[0].Digest
+		cumulativeTci := tcbinfo.Fwids[1].Digest
+		hasher.Reset()
+		hasher.Write(lastCumulativeTCI)
+		hasher.Write(currentTci)
+		expectedCumulativeValue := hasher.Sum(nil)
+		if !bytes.Equal(cumulativeTci, expectedCumulativeValue) {
+			return fmt.Errorf("cumulative TCI value for TCB block-%d, want %v but got %v", i, expectedCumulativeValue, cumulativeTci)
+		}
+		lastCumulativeTCI = cumulativeTci
+	}
```

Comment:
TCI_CUMULATIVE isn't a hash of all the currents from TcbInfos in the cert. TCI_CUMULATIVE gets extended whenever the ExtendTCI command is called on a node so for a given TCI Node X, TCI_CUMULATIVE is calculated using all the values `input` for each call to `ExtendTCI(handle=X, input_data=input)`.

Reply:
Rectified the issue